### PR TITLE
ignore remote update younger than recent ownership bid

### PR DESCRIPTION
**Pull request by AndrewMeadows**
* fix race condition between recent simulation ownership bid and on-the-wire update from remote owner

### CODE REVIEW

File: libraries/entities/src/EntityItem.cpp [hunk 1]
```diff
@@ -386,7 +386,13 @@ int EntityItem::readEntityDataFromBuffer(const unsigned char* data, int bytesLef
...
     }
 
-    qint64 clockSkew = args.sourceNode ? args.sourceNode->getClockSkewUsec() : 0;
+    int64_t clockSkew = 0;
+    uint64_t maxPingRoundTrip = 33333; // two frames periods at 60 fps
+    if (args.sourceNode) {
+        clockSkew = args.sourceNode->getClockSkewUsec();
+        const float MSECS_PER_USEC = 1000;
+        maxPingRoundTrip += args.sourceNode->getPingMs() * MSECS_PER_USEC;
+    }
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItem.cpp]: We assume a base max delay of 2 * 1/60 second, then add the real measured ping between the interface and the entity-server, and then use `maxPingRoundTrip` below to determine if we should ignore this update in favor of a pending ownership bid.

File: libraries/entities/src/EntityItem.cpp [hunk 2]
```diff
@@ -690,16 +695,19 @@ int EntityItem::readEntityDataFromBuffer(const unsigned char* data, int bytesLef
                 // recompute weOwnSimulation for later
                 weOwnSimulation = _simulationOwner.matchesValidID(myNodeID);
             }
-        } else if (newSimOwner.getID().isNull() && _simulationOwner.pendingTake(lastEditedFromBufferAdjusted)) {
-            // entity-server is trying to clear someone  else's ownership
-            // but we want to own it, therefore we ignore this clear event
-            // and pretend that we own it (we assume we'll get it soon)
+        } else if (_simulationOwner.pendingTake(now - maxPingRoundTrip)) {
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItem.cpp]: `SimulationOwner::pendingTake()` was already implemented.  We pass it a timestamp from the past: the latest possible time when a pending "take ownership" bid (if any exists) would still be valid.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/733/HighFidelity-Beta-PR10696-8d19395b0a78aae5b448f62cdf20fc19f7f6ce8b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/733/HighFidelity-Beta-PR10696-8d19395b0a78aae5b448f62cdf20fc19f7f6ce8b.dmg

**Comment by AndrewMeadows** [comment]
This PR is an effort to fix the "_ping-pong ball passes through the paddle_" problem in Hifi.  While this PR does NOT solve the problem 100% (sometimes the ping-pong ball can still pass through the paddle due to how the paddle is moved in the physics engine) it does provide an improvement.

## Test Plan

(1) Login UserA and run the script below.  It throws a new ball every 2 seconds.

(2) Login UserB and stand near the path of the balls, far enough away to see them coming.

(3) UserB enable visualization of the simulation ownership (Developer --> Physics --> Highlight Simulation Ownership).  Note that the left most icon shows the ownership status: red for "owned by someone else) and blue for "owned by me".

(4) Make  a dynamic box with `dimensions = { x: 0.04, y: 0.5, z: 0.5 }` but no gravity

(5) UserB grab the box like it is a large paddle and start whacking the balls as they come near your avatar.
--> on hifi/master the balls will tunnel through the paddle more often than in this PR, and will land on the floor behind the batter (UserA).
--> on this PR the balls may still get through the paddle, but they are less likely to do so and when they do they tend to be deflected by the paddle rather than passing straight through unscathed.

**Comment by CainFoool** [comment]
Hey @AndrewMeadows - is this PR ready for testing? If so, I'll be happy to take a look.

**Comment by YuppyPlays** [comment]
Testing with @CainFoool.

**Comment by YuppyPlays** [comment]
Testing complete:

[Myself as User A on Desktop, Cain in HMD as User B]

(1) Login UserA and run the script below. It throws a new ball every 2 seconds.
**Pass**
(2) Login UserB and stand near the path of the balls, far enough away to see them coming.
**Pass**
(3) UserB enable visualization of the simulation ownership (Developer --> Physics --> Highlight Simulation Ownership). Note that the left most icon shows the ownership status: red for "owned by someone else) and blue for "owned by me".
**Pass** Although I do not understand the requirement as there is no reference after this as to what to be looking out for?
(4) Make a dynamic box with dimensions = { x: 0.04, y: 0.5, z: 0.5 } but no gravity
**Pass** We found that the box of this size more often than not went through the box, however when doubled in size it was 100% accurate as to the whacking of the balls.
(5) UserB grab the box like it is a large paddle and start whacking the balls as they come near your avatar.
--> on hifi/master the balls will tunnel through the paddle more often than in this PR, and will land on the floor behind the batter (UserA).
--> on this PR the balls may still get through the paddle, but they are less likely to do so and when they do they tend to be deflected by the paddle rather than passing straight through unscathed.
**Pass**  The balls seem to not go through but only with a larger paddle box, when size specified worked occasionally but more often went through.

File: libraries/physics/src/ObjectActionTractor.cpp [hunk 3]
```diff
@@ -126,9 +126,18 @@ bool ObjectActionTractor::prepareForTractorUpdate(btScalar deltaTimeStep) {
...
+                    _lastPositionTarget = position;
+                }
                 _positionalTarget = position;
-                _linearVelocityTarget = linearVelocity;
+                if (deltaTimeStep > EPSILON) {
+                    // blend the new velocity with the old (low-pass filter)
+                    glm::vec3 newVelocity = (1.0f / deltaTimeStep) * (position - _lastPositionTarget);
+                    const float blend = 0.25f;
+                    _linearVelocityTarget = (1.0f - blend) * _linearVelocityTarget + blend * newVelocity;
+                }
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionTractor.cpp]: We measure the velocity of the target position (and filter it a bit by blending it with the old value).

File: libraries/physics/src/ObjectActionTractor.cpp [hunk 4]
```diff
@@ -169,19 +178,19 @@ void ObjectActionTractor::updateActionWorker(btScalar deltaTimeStep) {
...
-                targetVelocity = (-speed / offsetLength) * offset;
+                offsetVelocity = (-speed / offsetLength) * offset;
                 if (speed > rigidBody->getLinearSleepingThreshold()) {
                     forceBodyNonStatic();
                     rigidBody->activate();
                 }
             }
             // this action is aggresively critically damped and defeats the current velocity
-            rigidBody->setLinearVelocity(targetVelocity);
+            rigidBody->setLinearVelocity(glmToBullet(_linearVelocityTarget) + offsetVelocity);
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionTractor.cpp]: We slam the velocity of the slaved object to be the velocity of the target plus an extra term that tries to reconcile the offset between where the object is and its target position.

**Comment by AndrewMeadows** [comment]
Sorry @YuppyPlays, the testing on this PR is a bit tricky.  We know that big fat objects don't allow tunneling.  We can't eliminate the tunneling of the ball through a thin box, but we are trying to minimize it.  So the test is basically:

(1) Create a box that sometimes suffers tunneling when used as a paddle
(2) Hit lots of balls with the box to try to get an idea as to whether the ball tunnels less (but still some non-zero unavoidable amount) in this PR than in hifi/master.

Unfortunately, you need to hit lots of balls to try to accumulate some statistics as to the failure rate of the paddle.

I should also note: for best results don't rotate the paddle when you hit the ball.  Try to swing the paddle without flicking your wrist.  A rotating paddle is more likely to allow the ball to tunnel through so any change in behavior would be harder to notice when rotating the paddle.

Basically the question is:  Does the paddle work better in this PR than in hifi/master?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/817/HighFidelity-Beta-PR10696-b105ad83380e6a7f5485b71bb2898521f47b85cc.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/817/HighFidelity-Beta-PR10696-b105ad83380e6a7f5485b71bb2898521f47b85cc.dmg

**Comment by YuppyPlays** [comment]
Testing w/ @CainFoool.

**Comment by YuppyPlays** [comment]
Testing complete:
[Myself in Desktop as User A + Cain as HMD as User B]

Master results: 6/25 (around ~20%)
PR results: 25/25 **Passed.** (100%)


Hope this helps :)
